core/keyserver/epochs.go

```diff
@@ -125,39 +124,53 @@ func (s *Server) ListMutations(ctx context.Context, in *pb.ListMutationsRequest)
 		glog.Errorf("ListMutations(): adminstorage.Read(%v): %v", in.DirectoryId, err)
 		return nil, status.Errorf(codes.Internal, "Cannot fetch directory info")
 	}
-
-	start, err := parseToken(in.PageToken)
+	meta, err := s.batches.ReadBatch(ctx, in.DirectoryId, in.Epoch)
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Internal, "ReadBatch(%v, %v): %v", in.DirectoryId, in.Epoch, err)
 	}
-	// Read mutations from the database.
-	max, entries, err := s.mutations.ReadPage(ctx, d.DirectoryID, in.GetEpoch(), start, in.GetPageSize())
+	rt, err := SourceMap(meta.Sources).ParseToken(in.PageToken)
 	if err != nil {
-		glog.Errorf("ListMutations(): mutations.ReadRange(%v, %v, %v, %v): %v", d.MapID, in.GetEpoch(), start, in.GetPageSize(), err)
+		return nil, status.Errorf(codes.InvalidArgument, "Failed parsing page_token: %v: %v", in.PageToken, err)
+	}
+
+	// Read PageSize + 1 messages from the log to see if there is another page.
+	high := meta.Sources[rt.ShardId].HighestWatermark
+	msgs, err := s.logs.ReadLog(ctx, d.DirectoryID, rt.ShardId, rt.LowWatermark, high, in.PageSize+1)
+	if err != nil {
+		glog.Errorf("ListMutations(): ReadLog(%v, log: %v/(%v, %v], batchSize: %v): %v",
+			d.DirectoryID, rt.ShardId, rt.LowWatermark, high, in.PageSize, err)
 		return nil, status.Error(codes.Internal, "Reading mutations range failed")
 	}
-	indexes := make([][]byte, 0, len(entries))
-	mutations := make([]*pb.MutationProof, 0, len(entries))
-	for _, e := range entries {
-		mutations = append(mutations, &pb.MutationProof{Mutation: e})
-		indexes = append(indexes, e.GetIndex())
+	more := len(msgs) == int(in.PageSize+1)
```

Comment:
Just to double check. Does `more` mean "there are more entries in the same shard"? So if `more==false` we will still possibly return a token pointing at the next shard?

Reply:
yes. I'll adjust the variable name to be clearer